ide/editor.completion/src/org/netbeans/spi/editor/completion/CompletionCollector.java

```diff
@@ -0,0 +1,565 @@
+            @NonNull
+            public Builder preselect(boolean preselect) {
+                this.preselect = preselect;
+                return this;
+            }
+
+            /**
+             * A string that should be used when comparing this completion with other
+             * completions. When {@code null} the label is used as the sort text.
+             */
+            @NonNull
+            public Builder sortText(@NonNull String sortText) {
+                this.sortText = sortText;
+                return this;
+            }
+
+            /**
+             * A string that should be used when filtering a set of completions.
+             * When {@code null} the label is used as the filter.
+             */
+            @NonNull
+            public Builder filterText(@NonNull String filterText) {
+                this.filterText = filterText;
+                return this;
+            }
+
+            /**
+             * A string that should be inserted into a document when selecting
+             * this completion. When {@code null} the label is used as the insert text.
+             */
+            @NonNull
+            public Builder insertText(@NonNull String insertText) {
+                this.insertText = insertText;
+                return this;
+            }
+
+            /**
+             * The format of the insert text. The format applies to both the
+             * {@code insertText} property and the {@code newText} property of a provided
+             * {@code textEdit}. If omitted defaults to {@link TextFormat#PlainText}.
+             */
+            @NonNull
+            public Builder insertTextFormat(@NonNull TextFormat insertTextFormat) {
+                this.insertTextFormat = insertTextFormat;
+                return this;
+            }
+
+            /**
+             * An edit which is applied to a document when selecting this completion.
+             * When an edit is provided the value of {@code insertText} is ignored.
+             * The range of the edit must be a single line range and it must
+             * contain the position at which completion has been requested.
+             */
+            @NonNull
+            public Builder textEdit(@NonNull TextEdit textEdit) {
+                this.textEdit = textEdit;
+                return this;
+            }
+
+            /**
+             * A list of additional text edits that are applied when selecting this
+             * completion. Edits must not overlap (including the same insert position)
+             * with the main edit nor with themselves.
+             * Additional text edits should be used to change text unrelated to the
+             * current cursor position (for example adding an import statement at the
+             * top of the file if the completion item will insert an unqualified type).
+             */
+            @NonNull
+            public Builder additionalTextEdits(@NonNull List<TextEdit> additionalTextEdits) {
```

Comment:
Consider making this cumulative. something like `addTextEdit(TextEdit)`

Reply:
Changed to `additionalTextEdits(@NonNull CompletableFuture<List<TextEdit>> additionalTextEdits)` to support lazy computation on `completion resolve` request.